### PR TITLE
Enable folding sections when run under Travis

### PR DESCRIPTION
**Pull request by cognifloyd**
Add output when molecule is running in travis-ci to make the logs easier to jump through.
Travis log folding is explained [here](https://github.com/travis-ci/travis-ci/issues/2285).

#### PR Type
- Feature Pull Request

### CODE REVIEW

**Comment by cognifloyd** [comment]
Test failures are unrelated to this change.

**Review by ssbarnea — CHANGES_REQUESTED** [review]
Nice feature, sadly is not yet a generic terminal feature. What worries me is that in case of exceptions, the end section will never be printed, likely causing other issues, especially in cases where multiple commands are executed and some would fail.

Can we maybe catch the exceptions print them and the done bit end and raise them back? 

I am also inclined to opt-in for an implementation using a decorator if possible? Using a decorator should make it much easier to wrap the start/end of it, don't you think?

**Comment by greg-hellings** [comment]
This bit concerns me, "While not officially documented, I think you can rely on this behavior in the foreseeable future." But that was written in 2014 and appears to still be valid. So it doesn't concern me that much.

Does it make sense to provide an output mechanism like this specifically just for one CI provider? Do others have similar functionality that we ought to incorporate?

I have no issues with the structure of the code, unless we think we're going to be asked or expected to expand this functionality to support a bunch of other providers.

**Review by greg-hellings — APPROVED** [review]

**Comment by cognifloyd** [comment]
> Can we maybe catch the exceptions print them and the done bit end and raise them back?
>
> I am also inclined to opt-in for an implementation using a decorator if possible? Using a decorator should make it much easier to wrap the start/end of it, don't you think?

I don't like requiring all of the commands to add an extra print function at the end. Doing this in a decorator would increase the complexity of the Base command subclasses even more.

What if I modified `execute_scenario` instead, wrapping the call to `execute_subcommand` in a `try`/`finally` block to print the end of the fold line. Or it could be a func wrapper around `execute_subcommand`. I think I'll try a functools wrapper around `execute_subcommand`.

**Comment by cognifloyd** [comment]
@ssbarnea Using a decorator seems to be much simpler. Thanks for the idea. What do you think?

@greg-hellings if we do get requests to add something like this for other CI runners (I haven't seen a feature like this anywhere else), then the new wrapper isolates the CI-specific code more so that it's easy to tell what it is doing.

**Comment by cognifloyd** [comment]
To test this, I just started upgrading a molecule scenario to use molecule 3.1(ish) - my upgrade is not complete, but it shows how the folding looks:
https://travis-ci.com/github/cognifloyd/community.mongodb/builds/187516426

Sadly, it is not very informative at the folds. I need to play with that to see if there's a way to describe the fold when it is collapsed.

**Comment by cognifloyd** [comment]
Ooh. That looks better.
https://travis-ci.com/github/cognifloyd/community.mongodb/builds/187516426

I used this chain of links to learn how to do that:
https://github.com/travis-ci/docs-travis-ci-com/issues/949
https://github.com/travis-ci/travis-rubies/blob/9f7962a881c55d32da7c76baefc58b89e3941d91/build.sh#L38-L44

**Comment by cognifloyd** [comment]
OK. So, it's bike-shedding time:

### Text

1. `Molecule Scenario: 'default' Action: 'converge'`
2. `Molecule - Scenario: 'default' - Action: 'converge'`
3. `Molecule | Scenario: 'default' | Action: 'converge'`
4. `Molecule --> Scenario: 'default' --> Action: 'converge'`
5. `--> Scenario: 'default' --> Action: 'converge'`

### Color

Yellow bold (`33;1m`): _Current color_
Selection_089

Yellow normal (`33m`):
Selection_088

White bold (`37;1m`):
Selection_084

White normal (`37m`):
Selection_085

Cyan bold (`36;1m`):
Selection_086

Cyan normal (`36m`): _oops. I was hovering when I took this screenshot_
Selection_087

**Comment by cognifloyd** [comment]
My top 3:

1. Cyan bold
2. White normal
3. Yellow bold

File: molecule/command/base.py [hunk 1]
```diff
@@ -122,6 +123,31 @@ def execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args=()
...
+    @functools.wraps(func)
+    def travis_wrapper(config, subcommand):
+        print(
+            (
+                "travis_fold:start:{0}.{1}"
+                "\033[33;1mMolecule Scenario: '{0}' Action: '{1}'\033[0m"
+            ).format(
+                config.scenario.name,
+                subcommand,
+            )
```
Comment (cognifloyd) [inline molecule/command/base.py:140]:
```suggestion
            (
                "travis_fold:start:{0}.{1}"
                "{2}Molecule Scenario: '{0}' Action: '{1}'{3}"
            ).format(
                config.scenario.name,
                subcommand,
                colorama.Style.Bold + colorama.Fore.CYAN,
                colorama.Style.RESET_ALL,
            )
```

We might also be able to use `colorama` like in logger.py (would need to test to make sure the escape sequence works in this context (colorama uses `\x1b` fwict but the example I found uses `\033`).

**Comment by ssbarnea** [comment]
I would not spend much time investing in colorama as I am already working on a big change that drops it and switching to https://github.com/willmcgugan/rich

**Comment by cognifloyd** [comment]
Yes, the colorama escape sequence works just fine for Travis. Feel free to replace it with rich when you make the switch.

I went with Bold Cyan with the text `Molecule --> Scenario: 'default' --> Action: 'converge'`:
Selection_090

So, unless anyone else wants to join in the bike-shedding, I think this is ready to merge.

**Comment by cognifloyd** [comment]
I squashed and rebased.

Plus I've also prepared a [2.22 backport](https://github.com/cognifloyd/molecule/tree/travis-folding-2.22). Will there be a 2.23 release at some point?

**Review by ssbarnea — CHANGES_REQUESTED** [review]
That change is far from ready, being broken and not tested:
```
$ molecule list                                                                                                                 [18:36:45]
--> Found config file /Users/ssbarnea/.config/molecule/config.yml
travis_fold:start:default.listMolecule --> Scenario: 'default' --> Action: 'list'
travis_fold:end:default.list
Traceback (most recent call last):
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/bin/molecule", line 33, in <module>
    sys.exit(load_entry_point('molecule', 'console_scripts', 'molecule')())
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/ssbarnea/c/a/molecule/molecule/command/list.py", line 111, in list
    statuses.extend(base.execute_subcommand(scenario.config, subcommand))
TypeError: 'NoneType' object is not iterable
$ molecule list                                                                                                                 [18:36:45]
--> Found config file /Users/ssbarnea/.config/molecule/config.yml
travis_fold:start:default.listMolecule --> Scenario: 'default' --> Action: 'list'
travis_fold:end:default.list
Traceback (most recent call last):
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/bin/molecule", line 33, in <module>
    sys.exit(load_entry_point('molecule', 'console_scripts', 'molecule')())
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/ssbarnea/c/a/molecule/molecule/command/list.py", line 111, in list
    statuses.extend(base.execute_subcommand(scenario.config, subcommand))
TypeError: 'NoneType' object is not iterable
```

File: molecule/command/base.py [hunk 2]
```diff
@@ -122,6 +124,33 @@ def execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args=()
...
+                "{2}Molecule --> Scenario: '{0}' --> Action: '{1}'{3}"
+            ).format(
+                config.scenario.name,
+                subcommand,
+                colorama.Style.BRIGHT + colorama.Fore.CYAN,
+                colorama.Style.RESET_ALL,
+            )
+        )
+        try:
+            return func(config, subcommand)
```
Comment (cognifloyd) [inline molecule/command/base.py:146]: I missed this return here. Added. Please try again.

Comment (cognifloyd) [reply molecule/command/base.py:146]: For background, I discovered that I missed this in the 2.22 branch, but didn't check to see if it was also required for 3.x. Oops.
See: https://github.com/cognifloyd/molecule/commit/6ddc9b2717c43a89bacf24ef7d4a04b498d0fe90

**Comment by cognifloyd** [comment]
Should I add a functional test that runs on travis to test this code? Or add a functional test that triggers this by setting the TRAVIS and CI vars?

**Comment by cognifloyd** [comment]
I parametrized the unit test for `execute_subcommand` so that it tests as if running on Travis as well.
Do you want any functional or other tests for this? I suppose we could run this and check that the prints happen, but I'm not sure that has much value.

**Comment by ssbarnea** [comment]
Please rebase it, I moved molecule/ under lib/molecule. Sorry for the inconvenience.

**Comment by ssbarnea** [comment]
Closing due to lack of updates. Ping me if you want to reopen it.

**Comment by cognifloyd** [comment]
revisiting now. Please reopen.

**Comment by cognifloyd** [comment]
Here's what it looks like with the latest `rich` formatted output instead of straight colorama.
image

**Comment by cognifloyd** [comment]
Re-submitted PR as #2967

**Pull request closed without merging**